src/circuit.rs

```diff
@@ -1,46 +1,666 @@
+        let mut layouter = SingleChipLayouter::new(cs)?;
+
+        // Load the Sinsemilla generator lookup table used by the whole circuit.
+        SinsemillaChip::<pallas::Affine, SinsemillaK, SinsemillaC>::load(
+            config.sinsemilla_config_1.clone(),
+            &mut layouter,
+        )?;
+
+        // Construct the ECC chip.
+        let ecc_chip = config.ecc_chip();
+
+        // Witness private inputs that are used across multiple checks.
+        let (v_old, v_new, rho_old, psi_old, cm_old, g_d_old, ak, nk) = {
+            let v_old = {
+                // Witness v_old.
+                let v_old_val = self
+                    .v_old
+                    .map(|value| pallas::Base::from_u64(value.inner()));
+                self.load_private(
+                    layouter.namespace(|| "witness v_old"),
+                    config.advices[0],
+                    v_old_val,
+                )?
+            };
+
+            let v_new = {
+                // Witness v_new.
+                let v_new_val = self
+                    .v_new
+                    .map(|value| pallas::Base::from_u64(value.inner()));
+                self.load_private(
+                    layouter.namespace(|| "witness v_new"),
+                    config.advices[0],
+                    v_new_val,
+                )?
+            };
+
+            // Witness psi_old
+            let psi_old = self.load_private(
+                layouter.namespace(|| "witness psi_old"),
+                config.advices[0],
+                self.psi_old,
+            )?;
+
+            // Witness rho_old
+            let rho_old = self.load_private(
+                layouter.namespace(|| "witness rho_old"),
+                config.advices[0],
+                self.rho_old.map(|rho| rho.0),
+            )?;
+
+            // Witness cm_old
+            let cm_old = Point::new(
+                config.ecc_chip(),
+                layouter.namespace(|| "cm_old"),
+                self.cm_old.as_ref().map(|cm| cm.to_affine()),
+            )?;
+
+            // Witness g_d_old
+            let g_d_old = Point::new(
+                config.ecc_chip(),
+                layouter.namespace(|| "gd_old"),
+                self.g_d_old.as_ref().map(|gd| gd.to_affine()),
+            )?;
+
+            // Witness ak.
+            let ak: Option<pallas::Point> = self.ak.as_ref().map(|ak| ak.into());
+            let ak = Point::new(
+                ecc_chip.clone(),
+                layouter.namespace(|| "ak"),
+                ak.map(|ak| ak.to_affine()),
+            )?;
+
+            // Witness nk.
+            let nk = self.load_private(
+                layouter.namespace(|| "witness nk"),
+                config.advices[0],
+                self.nk.map(|nk| *nk),
+            )?;
+
+            (v_old, v_new, rho_old, psi_old, cm_old, g_d_old, ak, nk)
+        };
+
+        // Merkle path validity check.
+        // TODO: constrain output to equal public input
```

Comment:
More precisely: require that either `v_old = 0`, or the computed anchor equals the public input anchor.